chpl/chpl-service/src/main/java/gov/healthit/chpl/changerequest/domain/service/ChangeRequestStatusService.java

```diff
@@ -0,0 +1,176 @@
+
+        return crStatusDAO.create(cr, crStatus);
+    }
+
+    public ChangeRequest updateChangeRequestStatus(ChangeRequest crFromCaller)
+            throws EntityRetrievalException {
+        ChangeRequest crFromDb = crDAO.get(crFromCaller.getId());
+
+        // Check for nulls - a Java 8 way to check a chain of objects for null
+        Long statusTypeIdFromDB = Optional.ofNullable(crFromDb)
+                .map(ChangeRequest::getCurrentStatus)
+                .map(ChangeRequestStatus::getChangeRequestStatusType)
+                .map(ChangeRequestStatusType::getId)
+                .orElse(null);
+        Long statusTypeIdFromCaller = Optional.ofNullable(crFromCaller)
+                .map(ChangeRequest::getCurrentStatus)
+                .map(ChangeRequestStatus::getChangeRequestStatusType)
+                .map(ChangeRequestStatusType::getId)
+                .orElse(null);
+
+        if (statusTypeIdFromDB != null && statusTypeIdFromCaller != null
+                && statusTypeIdFromDB != statusTypeIdFromCaller
+                && isStatusChangeValid(statusTypeIdFromDB, statusTypeIdFromCaller)) {
+
+            createNewStatusForChangeRequest(
+                    crFromDb,
+                    statusTypeIdFromCaller,
+                    crFromCaller.getCurrentStatus().getComment());
+
+            // Need the updated CR with the new status
+            ChangeRequest updatedCrFromDB = crDAO.get(crFromDb.getId());
+
+            // Run any post processing based on the change request type
+            crDetailsFactory.get(updatedCrFromDB.getChangeRequestType().getId())
+                    .postStatusChangeProcessing(updatedCrFromDB);
+
+            try {
+                activityManager.addActivity(
+                        ActivityConcept.CHANGE_REQUEST,
+                        updatedCrFromDB.getId(),
+                        "Change request status updated",
+                        crFromDb,
+                        updatedCrFromDB);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+            return updatedCrFromDB;
+        } else {
+            return null;
+        }
+    }
+
+    public static boolean doesCurrentStatusExist(ChangeRequest cr) {
+        return cr.getCurrentStatus() != null
+                && cr.getCurrentStatus().getChangeRequestStatusType() != null
+                && cr.getCurrentStatus().getChangeRequestStatusType().getId() != null;
+    }
+
+    private boolean isStatusChangeValid(final Long previousStatusTypeId, final Long newStatusTypeId) {
+        // Does this status type id exist?
+        try {
+            crStatusTypeDAO.getChangeRequestStatusTypeById(newStatusTypeId);
+        } catch (EntityRetrievalException e) {
+            return false;
+        }
+
+        // Cannot change status is Cancelled
+        if (previousStatusTypeId == this.cancelledByRequesterStatus) {
```

Comment:
I think with "L"ong (as opposed to "l"ong) you should use .equals instead of ==. Not 100% on that but I feel like I've been burned before by using ==.

Reply:
I remember checkstyle reporting me as doing a bad thing by not using the `.equals()` in that case too